File: src/rollback/RollbackManager.js
'use strict';

const fs = require('fs');
const path = require('path');
const crypto = require('crypto');

/**
 * Manages rollback snapshots for destructive server operations.
 * Before any destructive action, a snapshot of relevant state is captured.
 * If something goes wrong, the snapshot can be restored.
 */
class RollbackManager {
  constructor(sshManager, options = {}) {
    this.ssh = sshManager;
    this.snapshotDir = options.snapshotDir || path.join(process.cwd(), '.sysadmin-snapshots');
    this.snapshots = new Map();

    // Ensure local snapshot directory exists
    if (!fs.existsSync(this.snapshotDir)) {
      fs.mkdirSync(this.snapshotDir, { recursive: true });
    }

    this._loadSnapshots();
  }

  /**
   * Create a rollback snapshot before a destructive action.
   * @param {string} command - The command about to be executed.
   * @param {string} description - Human-readable description.
   * @param {object} options - Additional snapshot options.
   * @returns {Promise<string>} The snapshot ID.
   */
  async createSnapshot(command, description, options = {}) {
    const id = crypto.randomUUID();
    const timestamp = new Date().toISOString();

    const snapshot = {
      id,
      timestamp,
      command,
      description,
      status: 'pending',
      backups: [],
    };

    // Determine what to back up based on the command
    const filesToBackup = this._identifyAffectedFiles(command);

    for (const remoteFile of filesToBackup) {
      try {
        const backupData = await this._backupRemoteFile(id, remoteFile);
        snapshot.backups.push(backupData);
      } catch (err) {
        snapshot.backups.push({ file: remoteFile, error: err.message });
      }
    }

    // Capture service states if the command involves services
    if (command.includes('systemctl') || command.includes('service')) {
      try {
        const { stdout } = await this.ssh.execute('systemctl list-units --type=service --state=running --no-pager --no-legend 2>/dev/null');
        snapshot.serviceState = stdout;
      } catch { /* ignore */ }
    }

    // Capture package state if command involves package management
    if (command.includes('apt') || command.includes('yum') || command.includes('dnf')) {
      try {
        const { stdout } = await this.ssh.execute(
          'dpkg --get-selections 2>/dev/null || rpm -qa 2>/dev/null'
        );
        snapshot.packageState = stdout;
      } catch { /* ignore */ }
    }

    snapshot.status = 'ready';
    this.snapshots.set(id, snapshot);
    this._saveSnapshots();

    return id;
  }

  /**
   * Rollback a specific snapshot - restore state to before the action.
   * @param {string} snapshotId - The snapshot to roll back.
   * @returns {Promise<object>} Rollback result.
   */
  async rollback(snapshotId) {
    const snapshot = this.snapshots.get(snapshotId);
    if (!snapshot) {
      throw new Error(`Snapshot ${snapshotId} not found`);
    }

    const results = [];

    // Restore backed-up files
    for (const backup of snapshot.backups) {
      if (backup.error) {
        results.push({ file: backup.file, status: 'skipped', reason: backup.error });
        continue;
      }

      try {
        await this._restoreRemoteFile(backup);
        results.push({ file: backup.file, status: 'restored' });
      } catch (err) {
        results.push({ file: backup.file, status: 'failed', error: err.message });
      }
    }

    // Restore service states if applicable
    if (snapshot.serviceState && snapshot.command.includes('systemctl')) {
      try {
        const serviceName = snapshot.command.match(/systemctl\s+\w+\s+(\S+)/);
        if (serviceName) {
          // Determine if the service was running before
          const wasRunning = snapshot.serviceState.includes(serviceName[1]);
          if (wasRunning) {
            await this.ssh.executeSudo(`systemctl start ${serviceName[1]}`);
          } else {
            await this.ssh.executeSudo(`systemctl stop ${serviceName[1]}`);
          }
          results.push({ service: serviceName[1], status: 'restored' });
        }
      } catch (err) {
        results.push({ service: 'unknown', status: 'failed', error: err.message });
      }
    }

    snapshot.status = 'rolled_back';
    snapshot.rollbackTime = new Date().toISOString();
    snapshot.rollbackResults = results;
    this._saveSnapshots();

    return { snapshotId, results };
  }

  /**
   * List all snapshots with their current status.
   */
  listSnapshots() {
    return Array.from(this.snapshots.values()).map((s) => ({
      id: s.id,
      timestamp: s.timestamp,
      command: s.command,
      description: s.description,
      status: s.status,
      backupCount: s.backups.length,
    }));
  }

  /**
   * Get details of a specific snapshot.
   */
  getSnapshot(snapshotId) {
    return this.snapshots.get(snapshotId) || null;
  }

  /**
   * Remove a snapshot (e.g., when action was not approved).
   */
  async removeSnapshot(snapshotId) {
    const snapshot = this.snapshots.get(snapshotId);
    if (!snapshot) return;

    // Clean up local backup files
    for (const backup of snapshot.backups) {
      if (backup.localPath && fs.existsSync(backup.localPath)) {
        fs.unlinkSync(backup.localPath);
      }
    }

    this.snapshots.delete(snapshotId);
    this._saveSnapshots();
  }

  /**
   * Identify files that may be affected by a command.
   */
  _identifyAffectedFiles(command) {
    const files = [];

    // Common config files affected by various commands
    const patterns = [
      { regex: /resolv\.conf/, file: '/etc/resolv.conf' },
      { regex: /hosts/, file: '/etc/hosts' },
      { regex: /sshd/, file: '/etc/ssh/sshd_config' },
      { regex: /nginx/, file: '/etc/nginx/nginx.conf' },
      { regex: /apache|httpd/, file: '/etc/apache2/apache2.conf' },
      { regex: /mysql|mariadb/, file: '/etc/mysql/my.cnf' },
      { regex: /php/, file: '/etc/php.ini' },
      { regex: /cron/, file: '/etc/crontab' },
      { regex: /fstab/, file: '/etc/fstab' },
      { regex: /iptables/, file: '/etc/iptables/rules.v4' },
    ];

    for (const pattern of patterns) {
      if (pattern.regex.test(command)) {
        files.push(pattern.file);
      }
    }

    // Extract explicit file paths from the command
    const filePathMatch = command.match(/(?:>|>>|cp|mv|rm|echo.*>)\s+(\S+)/g);
    if (filePathMatch) {
      for (const match of filePathMatch) {
        const filePath = match.replace(/^(>|>>|cp|mv|rm|echo.*>)\s+/, '').trim();
        if (filePath.startsWith('/')) {
          files.push(filePath);
        }
      }
    }

    return [...new Set(files)];
  }

  async _backupRemoteFile(snapshotId, remotePath) {
    const localDir = path.join(this.snapshotDir, snapshotId);
    if (!fs.existsSync(localDir)) {
      fs.mkdirSync(localDir, { recursive: true });
    }

    const safeFileName = remotePath.replace(/\//g, '_');
    const localPath = path.join(localDir, safeFileName);

    try {
      // Check if file exists on remote first
      const { code } = await this.ssh.execute(`test -f ${remotePath} && echo exists`);
      if (code !== 0) {
        return { file: remotePath, existed: false, localPath: null };
      }

      await this.ssh.downloadFile(remotePath, localPath);

      // Also capture file permissions
      const { stdout: perms } = await this.ssh.execute(`stat -c '%a %U %G' ${remotePath} 2>/dev/null`);

      return {
        file: remotePath,
        localPath,
        existed: true,
        permissions: perms.trim(),
      };
    } catch (err) {
      return { file: remotePath, error: err.message };
    }
  }

  async _restoreRemoteFile(backup) {
    if (!backup.existed) {
      // File didn't exist before, remove it if it was created
      await this.ssh.executeSudo(`rm -f ${backup.file}`);
      return;
    }

    if (!backup.localPath || !fs.existsSync(backup.localPath)) {
      throw new Error(`Local backup file not found: ${backup.localPath}`);
    }

    // Upload the backup
    const tempPath = `/tmp/.sysadmin-restore-${Date.now()}`;
    await this.ssh.uploadFile(backup.localPath, tempPath);
    await this.ssh.executeSudo(`cp ${tempPath} ${backup.file}`);

    // Restore permissions
    if (backup.permissions) {
      const [mode, owner, group] = backup.permissions.split(' ');
      await this.ssh.executeSudo(`chmod ${mode} ${backup.file}`);
      await this.ssh.executeSudo(`chown ${owner}:${group} ${backup.file}`);
    }

    // Clean up temp file
    await this.ssh.execute(`rm -f ${tempPath}`);
  }

  _saveSnapshots() {
    const dataFile = path.join(this.snapshotDir, 'snapshots.json');
    const data = Object.fromEntries(this.snapshots);
    fs.writeFileSync(dataFile, JSON.stringify(data, null, 2));
  }

  _loadSnapshots() {
    const dataFile = path.join(this.snapshotDir, 'snapshots.json');
    if (fs.existsSync(dataFile)) {
      try {
        const data = JSON.parse(fs.readFileSync(dataFile, 'utf-8'));
        for (const [id, snapshot] of Object.entries(data)) {
          this.snapshots.set(id, snapshot);
        }
      } catch { /* ignore corrupt file */ }
    }
  }
}

module.exports = RollbackManager;

File: sysadmin_agent/rollback/rollback_manager.py
import json
import re
import shutil
import uuid
from datetime import datetime, timezone
from pathlib import Path


COMMON_CONFIG_PATTERNS = [
    r"/etc/[\w./-]+",
    r"/var/[\w./-]+\.conf",
    r"/opt/[\w./-]+\.conf",
    r"~?/\.[\w./-]+",
]

REDIRECT_PATTERN = re.compile(r">{1,2}\s*(\S+)")
FILE_CMD_PATTERN = re.compile(r"\b(?:cp|mv|rm|install)\s+.*?(\S+)\s*$")
ECHO_TEE_PATTERN = re.compile(r"\btee\s+(?:-a\s+)?(\S+)")
SYSTEMCTL_PATTERN = re.compile(r"\bsystemctl\s+\w+\s+([\w@.-]+)")
PACKAGE_PATTERN = re.compile(r"\b(?:apt|apt-get|yum|dnf)\s+(?:install|remove|purge|update|upgrade)\b")


class RollbackManager:
    def __init__(self, ssh_manager, snapshot_dir=".sysadmin-snapshots"):
        self._ssh = ssh_manager
        self._snapshot_dir = Path(snapshot_dir)
        self._snapshot_dir.mkdir(parents=True, exist_ok=True)
        self._registry_path = self._snapshot_dir / "snapshots.json"
        self._snapshots = self._load_registry()

    def create_snapshot(self, command, description=""):
        snapshot_id = str(uuid.uuid4())
        snapshot_path = self._snapshot_dir / snapshot_id
        snapshot_path.mkdir(parents=True, exist_ok=True)

        affected_files = self._analyze_affected_files(command)
        backed_up_files = []

        for remote_path in affected_files:
            try:
                safe_name = remote_path.replace("/", "_").lstrip("_")
                local_path = snapshot_path / safe_name
                self._ssh.download_file(remote_path, str(local_path))

                # Capture permissions
                result = self._ssh.execute(f"stat -c '%a %U %G' {remote_path}")
                perms = result["stdout"].strip() if result["exit_code"] == 0 else None

                backed_up_files.append({
                    "remote_path": remote_path,
                    "local_file": safe_name,
                    "permissions": perms,
                })
            except Exception:
                continue

        service_states = self._capture_service_state(command)
        package_state = self._capture_package_state(command)

        metadata = {
            "id": snapshot_id,
            "command": command,
            "description": description,
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "status": "active",
            "files": backed_up_files,
            "service_states": service_states,
            "package_state": package_state,
        }

        self._snapshots[snapshot_id] = metadata
        self._save_registry()
        return snapshot_id

    def rollback(self, snapshot_id):
        snapshot = self._snapshots.get(snapshot_id)
        if not snapshot:
            raise ValueError(f"Snapshot {snapshot_id} not found")

        results = []
        snapshot_path = self._snapshot_dir / snapshot_id

        for file_info in snapshot.get("files", []):
            remote_path = file_info["remote_path"]
            local_file = snapshot_path / file_info["local_file"]

            if not local_file.exists():
                results.append({"file": remote_path, "status": "missing_backup"})
                continue

            try:
                self._ssh.upload_file(str(local_file), remote_path)

                if file_info.get("permissions"):
                    parts = file_info["permissions"].split()
                    if len(parts) == 3:
                        mode, owner, group = parts
                        self._ssh.execute(f"chmod {mode} {remote_path}")
                        self._ssh.execute(f"chown {owner}:{group} {remote_path}")

                results.append({"file": remote_path, "status": "restored"})
            except Exception as e:
                results.append({"file": remote_path, "status": "error", "error": str(e)})

        for svc in snapshot.get("service_states", []):
            try:
                name = svc["name"]
                was_active = svc.get("active", False)
                target_action = "start" if was_active else "stop"
                self._ssh.execute(f"systemctl {target_action} {name}")
                results.append({"service": name, "status": f"{target_action}ed"})
            except Exception as e:
                results.append({"service": svc.get("name"), "status": "error", "error": str(e)})

        snapshot["status"] = "rolled_back"
        self._save_registry()
        return results

    def list_snapshots(self):
        return [
            {
                "id": s["id"],
                "command": s["command"],
                "description": s["description"],
                "timestamp": s["timestamp"],
                "status": s["status"],
                "file_count": len(s.get("files", [])),
            }
            for s in self._snapshots.values()
        ]

    def get_snapshot(self, snapshot_id):
        return self._snapshots.get(snapshot_id)

    def remove_snapshot(self, snapshot_id):
        if snapshot_id not in self._snapshots:
            return False

        snapshot_path = self._snapshot_dir / snapshot_id
        if snapshot_path.exists():
            shutil.rmtree(snapshot_path)

        del self._snapshots[snapshot_id]
        self._save_registry()
        return True

    def _analyze_affected_files(self, command):
        files = set()

        for pattern in COMMON_CONFIG_PATTERNS:
            for match in re.finditer(pattern, command):
                files.add(match.group(0))

        for match in REDIRECT_PATTERN.finditer(command):
            path = match.group(1)
            if path.startswith("/"):
                files.add(path)

        for match in ECHO_TEE_PATTERN.finditer(command):
            path = match.group(1)
            if path.startswith("/"):
                files.add(path)

        for match in FILE_CMD_PATTERN.finditer(command):
            path = match.group(1)
            if path.startswith("/"):
                files.add(path)

        return sorted(files)

    def _capture_service_state(self, command):
        states = []
        for match in SYSTEMCTL_PATTERN.finditer(command):
            service_name = match.group(1)
            try:
                result = self._ssh.execute(f"systemctl is-active {service_name}")
                active = result["stdout"].strip() == "active"
                states.append({"name": service_name, "active": active})
            except Exception:
                states.append({"name": service_name, "active": None})
        return states

    def _capture_package_state(self, command):
        if not PACKAGE_PATTERN.search(command):
            return None

        try:
            result = self._ssh.execute("dpkg --get-selections 2>/dev/null || rpm -qa 2>/dev/null")
            if result["exit_code"] == 0:
                return result["stdout"]
        except Exception:
            pass
        return None

    def _load_registry(self):
        if self._registry_path.exists():
            try:
                data = json.loads(self._registry_path.read_text())
                return {s["id"]: s for s in data}
            except (json.JSONDecodeError, KeyError):
                return {}
        return {}

    def _save_registry(self):
        data = list(self._snapshots.values())
        self._registry_path.write_text(json.dumps(data, indent=2))
